File: Benchmarking/stats/euler112/euler112.jl
#=
Working from left-to-right if no digit is exceeded by the digit to its left it is called an increasing number; for example, 134468.

Similarly if no digit is exceeded by the digit to its right it is called a decreasing number; for example, 66420.

We shall call a positive integer that is neither increasing nor decreasing a "bouncy" number; for example, 155349.

Clearly there cannot be any bouncy numbers below one-hundred, but just over half of the numbers below one-thousand (525) are bouncy. In fact, the least number for which the proportion of bouncy numbers first reaches 50% is 538.

Surprisingly, bouncy numbers become more and more common and by the time we reach 21780 the proportion of bouncy numbers is equal to 90%.

Find the least number for which the proportion of bouncy numbers is exactly 99%.
=#
#Answer: 1587000

#Terminal: julia euler11.jl 99.00
prop = parse(Float64, ARGS[1])
#num = parse(Int, ARGS[2])

function kind(n)
  #Returns -1 for decreasing
  #Returns 0 for bouncy
  #Returns 1 for increasing
  s = string(n)
  l = length(s)
  i = 1
  j = i + 1
  decr, bouncy, incr = false, false, false
  while j <= l
    a = parse(Int8, s[i])
    b = parse(Int8, s[j])
    if s[i] > s[j]
      decr = true
    elseif s[i] < s[j]
      incr = true
    end
    i += 1
    j += 1
  end
  if decr && incr
    return true #Number is bouncy
  end
  return false
end

function calc(prop)
  currentProp = Float64(0)
  i = UInt128(100) #Starting number
  countBouncy = 0
  while currentProp < prop
    if kind(i)
      countBouncy += 1
    end

    currentProp = (countBouncy * 100) / i

    if currentProp >= prop
      return i
    end
    i += 1
  end
  
  return "Proportion was not reached."
end

calc(prop)

File: Benchmarking/stats/euler112/euler112.cpp
//
//  main.cpp
//  Euler112
//
//  Created by Sonni on 22/05/16.
//  Copyright © 2016 Sonni Jensen. All rights reserved.
//

#include <iostream>

bool kind( int n)
{
    std::string s = std::to_string(n);
    unsigned int l = s.size();
    unsigned int i = 1;
    unsigned int j = i + 1;
    bool decr = false, bouncy = false, incr = false;
    
    while (j < l)
    {
        unsigned int a = (int) s[i];
        unsigned int b = (int) s[j];
        if (a > b)
            decr = true;
        else if (a < b)
            incr = true;
        
        i++;
        j++;
    }
    if (decr && incr)
        return true;
    
    return false;
}

long long calc(long long prop)
{
    int currentProb = 0;
    
    long long i = 100;
    int countBouncy = 0;
    
    while (currentProb < prop)
    {
        if (kind(i))
            countBouncy++;
        
        currentProb = (countBouncy * 100) / i;
        
        if (currentProb >= prop)
            return i;
        i++;
    }
    
    return -1;
}

int main(int argc, const char * argv[]) {
    calc(atoi(argv[1]));
    return 0;
}
